From 953b6c44e30a502fd63b285da3cf5f827f7d604f Mon Sep 17 00:00:00 2001
From: Russell Brown <russell.brown@joyent.com>
Date: Thu, 26 Oct 2017 20:14:03 +0000
Subject: [PATCH] MANTA-2412 Allow addition of content-disposition header to
 manta objects

Add external dep for validating/cleaning content-disposition header
Add util fun for validiting content-disposition
Add tests for content-disposition
---
 lib/common.js           |   1 +
 lib/obj.js              |   2 +
 lib/server.js           |  16 ++++-
 lib/uploads/create.js   |  15 ++++-
 lib/utils.js            |  43 +++++++++++++-
 package.json            |   3 +-
 test/helper.js          |  10 ++--
 test/mpu/create.test.js |  47 ++++++++++++++-
 test/obj.test.js        | 127 ++++++++++++++++++++++++++++++++++++++++
 9 files changed, 252 insertions(+), 12 deletions(-)

diff --git a/lib/common.js b/lib/common.js
index 4d99268..ed92245 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -229,6 +229,7 @@ function createMetadata(req, type, cb) {
         break;
 
     case 'object':
+        md.contentDisposition = req.header('content-disposition');
         md.contentLength = req._size !== undefined ?
             req._size : prev.contentLength;
         md.contentMD5 = req._contentMD5 || prev.contentMD5;
diff --git a/lib/obj.js b/lib/obj.js
index 7d82bb9..14a94d8 100644
--- a/lib/obj.js
+++ b/lib/obj.js
@@ -727,6 +727,7 @@ function streamFromSharks(req, res, next) {
     if (md.contentLength === 0 || req.method === 'HEAD') {
         log.debug('streamFromSharks: HEAD || zero-byte object');
         res.header('Durability-Level', req.metadata.sharks.length);
+        res.header('Content-Disposition', req.metadata.contentDisposition);
         res.header('Content-Length', md.contentLength);
         res.header('Content-MD5', md.contentMD5);
         res.header('Content-Type', md.contentType);
@@ -750,6 +751,7 @@ function streamFromSharks(req, res, next) {
             res.header('Content-MD5', md.contentMD5);
         }
 
+        res.header('Content-Disposition', req.metadata.contentDisposition);
         res.header('Content-Length', sh['content-length']);
         res.header('Durability-Level', req.metadata.sharks.length);
 
diff --git a/lib/server.js b/lib/server.js
index c23229d..05983e1 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 var crypto = require('crypto');
@@ -32,6 +32,8 @@ var picker = require('./picker');
 var uploads = require('./uploads');
 var throttle = require('./throttle');
 
+var muskieUtils = require('./utils');
+
 // injects into the global namespace
 require('./errors');
 
@@ -198,6 +200,18 @@ function createServer(options, clients, name) {
         next();
     });
 
+    server.pre(function checkContentDisposition(req, res, next) {
+        muskieUtils.validateContentDisposition(
+            req.headers, function cb(err, _h) {
+                if (err) {
+                    req.log.debug('malformed content-disposition: %s', err.msg);
+                    res.send(400);
+                    next(false);
+                }
+                next();
+            });
+    });
+
     server.pre(function routeLiveJobs(req, res, next) {
         var tmp = JOBS_ROOT_RE.exec(req.path());
         if (tmp && tmp.length > 1 && JOBS_LIVE_RE.test(req.query()))
diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 9f7b46c..b1b1cc5 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -30,6 +30,7 @@ var auth = require('../auth');
 var common = require('../common');
 var obj = require('../obj');
 var uploadsCommon = require('./common');
+var muskieUtils = require('../utils');
 require('../errors');
 
 
@@ -178,6 +179,7 @@ function validateSchema(req, res, next) {
  *   - durability-level
  *   - content-length
  *   - content-md5
+ *   - content-disposition
  *
  * This handler is expected to set the following state on the upload object:
  * - objectPath
@@ -243,6 +245,17 @@ function validateParams(req, res, next) {
                 return;
             }
 
+            // validate the content-disposition header
+            muskieUtils.validateContentDisposition(
+                headers, function cb(cdErr, _h) {
+                    if (cdErr) {
+                        req.log.debug('malformed content-disposition: %s',
+                                      cdErr.msg);
+                        res.send(400);
+                        next(false);
+                    }
+                });
+
             assert.string(p);
             assert.object(headers);
             assert.number(size);
diff --git a/lib/utils.js b/lib/utils.js
index 434ac7e..8678164 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -5,9 +5,12 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
+var assert = require('assert-plus');
+var contentDisposition = require('content-disposition');
+
 //
 // Given a list of lists, this function will interleave each element
 // from each list into a new flat list.
@@ -72,11 +75,45 @@ function shuffle(array) {
 }
 
 
+/**
+ * validates, and sets a content-disposition header, if present
+ *
+ * If a content-dispositon key is present in the passed headers
+ * object, it is validated and canonicalised. The canonical, valid
+ * value is saved against the headers 'content-disposition' key and
+ * callback called with headers. If the value is invalid, an error is
+ * passed to the callback, the error.msg will be the invalid
+ * content-disposition string.
+ */
+function validateContentDisposition(headers, cb) {
+    assert.object(headers, 'headers');
+    assert.func(cb, 'callback');
+
+    if (headers['content-disposition'] !== undefined) {
+        var cd = headers['content-disposition'];
+
+        try {
+            // use a round trip of parse and write to set a valid
+            // value that can be used when reading the object.
+            var cdp = contentDisposition.parse(cd);
+            var cdStr = contentDisposition(cdp.parameters.filename,
+                                           cdp.type);
+            headers['content-disposition'] = cdStr;
+            cb(null, headers);
+        } catch (err) {
+            err.msg = cd;
+            cb(err, headers);
+        }
+    } else {
+        cb(null, headers);
+    }
+}
 
 ///--- Exports
 
 module.exports = {
     interleave: interleave,
     shallowCopy: shallowCopy,
-    shuffle: shuffle
-};
\ No newline at end of file
+    shuffle: shuffle,
+    validateContentDisposition: validateContentDisposition
+};
diff --git a/package.json b/package.json
index 3e07661..de2a237 100644
--- a/package.json
+++ b/package.json
@@ -18,6 +18,7 @@
         "bunyan": "0.22.1",
         "bunyan-syslog": "0.2.2",
         "cmdln": "4.3.0",
+        "content-disposition": "0.5.3",
         "cueball": "2.2.9",
         "dashdash": "1.3.2",
         "deep-equal": "0.0.0",
@@ -27,7 +28,7 @@
         "kang": "1.1.0",
         "keep-alive-agent": "0.0.1",
         "keyapi": "git+https://github.com/joyent/keyapi.git#e14b3d58",
-        "libmanta": "git+https://github.com/joyent/node-libmanta.git#v1.2.0",
+        "libmanta": "file:../node-libmanta",
         "libuuid": "0.1.2",
         "lru-cache": "4.1.5",
         "lstream": "0.0.4",
diff --git a/test/helper.js b/test/helper.js
index 1f183ee..1ea15d8 100644
--- a/test/helper.js
+++ b/test/helper.js
@@ -249,11 +249,11 @@ function checkResponse(t, res, code) {
     if (!res)
         return;
     t.equal(res.statusCode, code, 'HTTP status code mismatch');
-    t.ok(res.headers);
-    t.ok(res.headers.date);
-    t.equal(res.headers.server, 'Manta');
-    t.ok(res.headers['x-request-id']);
-    t.ok(res.headers['x-server-name']);
+    t.ok(res.headers, 'has headers');
+    t.ok(res.headers.date, 'headers have date');
+    t.equal(res.headers.server, 'Manta', 'server header is Manta');
+    t.ok(res.headers['x-request-id'], 'headers have x-req-id');
+    t.ok(res.headers['x-server-name'], 'headers have x-server-name');
 
     if (code === 200 || code === 201 || code === 202) {
         t.ok(res.headers['content-type']);
diff --git a/test/mpu/create.test.js b/test/mpu/create.test.js
index a1d752b..c1b50e0 100644
--- a/test/mpu/create.test.js
+++ b/test/mpu/create.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 var path = require('path');
@@ -464,3 +464,48 @@ test('create upload: x-durability-level less than allowed', function (t) {
         t.end();
     });
 });
+
+// content-disposition
+test('create upload: content-disposition header', function (t) {
+    var self = this;
+    var cd = 'attachment; filename="my-file.txt"';
+    var h = {
+        'content-disposition': cd
+    };
+
+    self.createUpload(self.path, h, function (err, o) {
+        if (ifErr(t, err, 'created upload')) {
+            t.end();
+            return;
+        }
+
+        self.getUpload(self.uploadId, function (err2, upload) {
+            if (ifErr(t, err2, 'got upload')) {
+                t.end();
+                return;
+            }
+
+            t.deepEqual(upload.headers, h, 'created headers match');
+            t.ok(upload.state, 'created');
+            t.end();
+        });
+    });
+});
+
+
+test('create upload: invalid content-disposition', function (t) {
+    var self = this;
+    var h = {
+        'content-disposition': 'attachment;'
+    };
+
+    self.createUpload(self.path, h, function (err, o) {
+        t.ok(err, 'Expect error');
+        if (!err) {
+            return (t.end());
+        }
+        t.ok(verror.hasCauseWithName(err,
+            'BadRequestError'), 'Expected 400');
+        t.end();
+    });
+});
diff --git a/test/obj.test.js b/test/obj.test.js
index c4863a1..61beb35 100644
--- a/test/obj.test.js
+++ b/test/obj.test.js
@@ -1058,3 +1058,130 @@ test('put timeout', function (t) {
         stream.write(TEXT.substr(0, 1));
     });
 });
+
+test('Put-Get content-disposition', function (t) {
+    var self = this;
+    var cd = 'attachment; filename="my-file.txt"';
+    var opts = {
+        headers: {
+            'content-disposition': cd
+        }
+    };
+
+    this.putObject(t, opts, function (_, headers) {
+        self.client.get(self.key, opts, function (err, stream, res) {
+            t.ifError(err);
+            t.checkResponse(res, 200);
+            t.equal(res.headers['content-disposition'], cd,
+                    'Content-Disposition should match written value');
+            t.end();
+        });
+    });
+});
+
+test('Put-Get content-disposition cleaned', function (t) {
+    var self = this;
+    var cd = 'attachment; filename="/root/my-file.txt"';
+    var opts = {
+        headers: {
+            'content-disposition': cd
+        }
+    };
+
+    this.putObject(t, opts, function (_, headers) {
+        self.client.get(self.key, opts, function (err, stream, res) {
+            t.ifError(err);
+            t.checkResponse(res, 200);
+            t.equal(res.headers['content-disposition'],
+                    'attachment; filename="my-file.txt"',
+                    'Content-Disposition should be clean');
+            t.end();
+        });
+    });
+});
+
+
+test('streaming object valid content-disposition',
+     function (t) {
+         var stream = new MemoryStream();
+         var text = 'The lazy brown fox \nsomething \nsomething foo';
+
+         process.nextTick(stream.end.bind(stream, text));
+         var opts = {
+             headers: {
+                 'content-disposition': 'attachment; filename="my-file.txt"'
+             }
+         };
+
+         this.client.put(this.key, stream, opts, function (err, res) {
+             t.ifError(err);
+             t.checkResponse(res, 204);
+             t.end();
+         });
+     });
+
+
+test('streaming object invalid content-disposition',
+     function (t) {
+         var stream = new MemoryStream();
+         var text = 'The lazy brown fox \nsomething \nsomething foo';
+
+         process.nextTick(stream.end.bind(stream, text));
+         var opts = {
+             headers: {
+                 'content-disposition': 'attachment;'
+             }
+         };
+         this.client.put(this.key, stream, opts, function (err, res) {
+             t.equal(res.statusCode, 400, 'Expected 400');
+             t.equal(err.name, 'BadRequestError', 'Expected a BadRequestError');
+             t.end();
+         });
+     });
+
+test('chattr: valid content-disposition', function (t) {
+    var cd = 'attachment; filename="my-file.txt"';
+    var opts = {
+        headers: {
+            'content-disposition': cd
+        }
+    };
+
+    var self = this;
+
+    this.putObject(t, function () {
+        self.client.chattr(self.key, opts, function (err) {
+            t.ifError(err);
+
+            self.client.info(self.key, function (err2, info) {
+                t.ifError(err2);
+                t.ok(info);
+                if (info) {
+                    var h = info.headers || {};
+                    t.equal(h['content-disposition'], cd,
+                            'Content-Disposition should match written value');
+                }
+                t.end();
+            });
+        });
+    });
+});
+
+test('chattr invalid content-disposition',
+     function (t) {
+         var opts = {
+             headers: {
+                 'content-disposition': 'attachment;'
+             }
+         };
+         var self = this;
+
+         this.putObject(t, function () {
+             self.client.chattr(self.key, opts, function (err, res) {
+                 t.equal(res.statusCode, 400, 'Expected 400');
+                 t.equal(err.name, 'BadRequestError',
+                         'Expected a BadRequestError');
+                 t.end();
+             });
+         });
+     });
-- 
2.21.0

